18-websockets/classworks/Gloria_Loi/web/src/pages/Chat/components/Chat/Chat.jsx

```diff
@@ -0,0 +1,132 @@
+import React, { useEffect, useState, useCallback } from 'react';
+import openSocket from 'socket.io-client';
+import { debounce } from 'lodash';
+
+import Message from '../Message';
+
+import './Chat.css';
+
+const socket = openSocket('http://localhost:8000');
+
+const Chat = () => {
+  const [nameInputValue, setNameInputValue] = useState('anonym');
+  const [messageInputValue, setMessageInputValue] = useState('');
+  const [userName, setUserName] = useState('anonym');
+  const [messages, setMessages] = useState([]);
+  const [typingUsers, setTypingUsers] = useState([]);
+
+  const handleNameChange = (value) => {
+    setUserName(value);
+  };
+
+  const sendMessage = (value) => {
+    socket.emit('new_message', { value, userName });
+  };
+
+  const handleTyping = () => {
+    socket.emit('typing', { userName });
+  };
+
+  const onMessage = useCallback(
+    (message) => {
+      console.log(message);
+      console.log(messages);
+      setMessages([...messages, message]);
+      console.log(messages);
+    },
+    [messages]
+  );
+
+  const onTyping = useCallback(
+    debounce(
+      (user) => {
+        clearTimeout(user.interval);
+        if (!typingUsers.includes(user.userName)) {
+          setTypingUsers([...typingUsers, user.userName]);
+        }
+        user.interval = setTimeout(() => {
+          setTypingUsers(
+            typingUsers.filter((typingUser) => typingUser !== user.userName)
+          );
+        }, 3000);
+        console.log(typingUsers);
+      },
+      [typingUsers],
+      2000,
+      { leading: true }
+    )
+  );
```

Comment:
`Actually, it doesn't work on my PC. Maybe `useCallback` is unnecessary. 
Also, I cannot understand why `user` has `interval` field. `user` is data from the server, why interval should be inside?